File: leetcode/lc-us/py/416.py
from typing import List


class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total_sum = sum(nums)
        if total_sum & 1: return False
        half_sum, dp = total_sum // 2, 1
        for num in nums:
            dp |= dp << num
        return dp & 1 << half_sum

File: leetcode/lc-us/cpp/416.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool canPartition(vector<int> &nums) {
        bitset<10001> f;
        int sum = 0;
        f[0] = true;
        for (int num: nums) {
            f |= f << num;
            sum += num;
        }
        if (sum & 1) return false;
        return f[sum / 2];
    }
};